napari/utils/colormaps/colormap.py

```diff
@@ -0,0 +1,46 @@
+import numpy as np
+from scipy import interpolate
+
+from .colorbars import make_colorbar
+from .standardize_color import transform_color
+from ...utils.dataclass import dataclass, Property
+
+
+@dataclass(events=True, properties=True)
+class Colormap:
+    """Colormap that relates intensity values to colors.
+
+    Parameters
+    ----------
+    name : str
+        Name of the colormap.
+    data : array, shape (N, 4)
+        Data used in the colormap.
+    controls : array, shape (N,) or (N+1,)
+        Control points of the colormap.
+    interpolation : str
+        Colormap interpolation mode, either `linear` or
+        `zero`.
+    """
+
+    name: str = 'undefined'
+    colors: Property[np.ndarray, None, transform_color] = np.zeros((2, 4))
+    controls: Property[np.ndarray, None, np.asarray] = np.asarray(
+        [0, 1]
+    )  # Not yet implemented
+    interpolation: str = 'linear'  # Only linear supported
```

Comment:
huh... that is quite nice :joy:

Reply:
we can reproduce all the functionality of all the colormaps we currently have (apart from the two 3D ones that I've never been such a fan of) without `controls` or `interpolation`.

I could see how they might be nice to have, and will be easy to implement, but we could also not

Reply:
For `controls` the thing I'm not so sure how to do with the dataclass is handle the init where None is provided and then it auto-calculates based on length of `colors` and `interpolation`. In general the dataclasses I find a little more tricky when there is coupling between the attributes.

Reply:
I've now added support for controls, interpolation, though i need to test `interpolation='zero'`. Also there are no custom setters now so things could fall out of place once made, but we might want to make this whole thing frozen ...